src/shared/sleep-config.c

```diff
@@ -102,6 +102,88 @@ int parse_sleep_config(SleepConfig **ret_sleep_config) {
         return 0;
 }
 
+bool is_battery_low(void) {
+        _cleanup_free_ char *bat_cap_level = NULL;
+        int r;
+
+        r = read_one_line_file("/sys/class/power_supply/BAT0/capacity_level", &bat_cap_level);
+        if (r < 0) {
+                log_debug_errno(r, "Failed to read /sys/class/power_supply/BAT0/capacity_level: %m");
+                return false;
+        }
+
+        log_debug("Current battery level: %s", bat_cap_level);
+
+        return STR_IN_SET(bat_cap_level, "Critical", "Low");
+}
+
+int read_battery_capacity(void) {
+        _cleanup_free_ char *bat_cap = NULL;
+        int battery_capacity, r;
+
+        r = read_one_line_file("/sys/class/power_supply/BAT0/capacity", &bat_cap);
+        if (r < 0)
+               return log_debug_errno(r, "Failed to read /sys/class/power_supply/BAT0/capacity: %m");
+
+        r = safe_atoi(bat_cap, &battery_capacity);
+        if (r < 0)
+               return log_debug_errno(r, "Failed to convert battery capacity char value to int: %m");
+
+        log_debug("Current battery charge level: %d%%", battery_capacity);
+
+        return battery_capacity;
+}
+
+int get_battery_discharge_rate(void) {
+        _cleanup_free_ char *dirpath = NULL;
+        int stored_discharge_rate, r;
+        char *dis_rate;
+        sd_id128_t machine_id = SD_ID128_NULL;
+
+        r = sd_id128_get_machine(&machine_id);
+        if (r < 0)
+                return log_error_errno(r, "Failed to get machine-id: %m");
```

Comment:
I have not follow the previous discussions yet, but why the file needs to be in the machine-id sub-directory??

Please use `log_debug_errno()`. This is a kind of library like function.

Reply:
> I have not follow the previous discussions yet, but why the file needs to be in the machine-id sub-directory??

To reduce the chances it gets used on a different installation (backup and restore after os reinstall or such), given it's very specific to a precise machine + installation

Reply:
> given it's very specific to a precise machine + installation

Theoretically yes, but is it significantly different?? I do not think so...

Reply:
Well yes, batteries have very different capacities and degrade in different ways between different machines. Even different installations might have different CPU governors and deep-sleep settings, affecting the discharge rate, so using the wrong value could mean unclean shutdown on power cut and lead to data loss.
What issue do you see in having the machine-id be part of this?

Reply:
Let me confirm, "machines" you mean hardware, right? The rate of course depends on the hardware. But the file is not shared with other hardware, right? Or, do you expect the file may be saved on a network file system??

If installation (especially, kernel and its settings e.g. CPU governor as you mentioned) heavily affects the rate, why not to clear the file before a new sleep operation is started, and always measure the discharge rate during the operation?? The kernel or setting may be changed after the last sleep, and in that case, the saved rate is not reliable.

> What issue do you see in having the machine-id be part of this?

Nothing, but it looks needlessly complicated. And I do not think it provides much safety and merits.